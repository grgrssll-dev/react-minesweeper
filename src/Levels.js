const Levels = [
	{
		name: 'Beginner',
		rows: 10,
		cols: 10,
		mines: 10
	},
	{
		name: 'Intermediate',
		rows: 16,
		cols: 16,
		mines: 40
	},
	{
		name: 'Expert',
		rows: 16,
		cols: 30,
		mines: 99
	}
];

export default Levels;
