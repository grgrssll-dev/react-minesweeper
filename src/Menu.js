import React, { PureComponent } from 'react';
import './Menu.css';

class Menu extends PureComponent {

	constructor(props) {
		super(props);
		this.state = {
			isMenuOpen: false
		};
	}

	_toggleMenu() {
		this.setState((state) => {
			return {
				isMenuOpen: !state.isMenuOpen
			};
		});
	}

	_closeMenu() {
		this.setState({
			isMenuOpen: false
		});
	}

	render() {
		console.log('-- menu render');
		const { levels, currentLevel, onLevelChange, onNewGame } = this.props;
		const { isMenuOpen } = this.state;
		return (
			<div className="menu">
				<ul className="menu--item">
					<li onClick={() => {
						this._toggleMenu();
					}}>Game</li>
					{isMenuOpen && (
						<ul className="menu--submenu">
							<li className="menu--divider-bottom"
								onClick={() => {
									this._closeMenu();
									onNewGame();
								}}>New</li>
							{Object.values(levels).map((level) => {
								return (
									<li key={level.name}
										onClick={() => {
											this._closeMenu();
											onLevelChange(level)
										}}
										className={level.name === currentLevel.name ? 'selected' : ''}>{level.name}</li>
								);
							})}
						</ul>
					)}
				</ul>
			</div>
		);
	}
}

export default Menu;
