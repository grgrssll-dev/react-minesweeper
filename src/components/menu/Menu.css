.menu {
	background: var(--system-bg);
	box-sizing: border-box;
	height: 1.5rem;
	padding: 0.375rem;
	position: relative;
	font-size: 0;
}

.menu--item {
	margin: 0;
	padding: 0;
	font-size: 0;
}

.menu--item > li {
	color: var(--system-text);
	cursor: default;
	display: inline-block;
	font-size: 0.75rem;
	line-height: 1;
	vertical-align: baseline;
}

.menu--submenu {
	background: var(--system-bg);
	border-bottom-color: var(--system-border-dark);
	border-left-color: var(--system-border-light);
	border-right-color: var(--system-border-dark);
	border-style: solid;
	border-top-color: var(--system-border-light);
	border-width: 1px;
	left: 0;
	margin: 0;
	min-width: 9rem;
	padding: 0;
	position: absolute;
	top: 100%;
}
.menu--submenu > li {
	box-sizing: border-box;
	cursor: default;
	font-size: 0.75rem;
	list-style-type: none;
	padding: 0.375rem 0.375rem 0.375rem 1.375rem;
	position: relative;
}

.menu--submenu > li.selected::before {
	content: 'âœ“';
	display: block;
	position: absolute;
	top: 0.375rem;
	left: 0.375rem;
}

.menu--divider-bottom::after {
	content: '';
	display: block;
	position: absolute;
	top: 100%;
	left: 0.125rem;
	width: calc(100% - 0.25rem);
	height: 0;
	border-top: 0.0625rem var(--system-border-dark) solid;
	border-bottom: 0.0625rem var(--system-border-light) solid;
}