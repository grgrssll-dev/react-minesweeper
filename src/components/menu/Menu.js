import React, { useState } from 'react';
import styled from '@emotion/styled';

function menu(props) {
	console.log('-- menu render');
	const [isMenuOpen, setMenuOpen] = useState(false);
	const { levels, currentLevel, onLevelChange, onNewGame } = props;

	const toggleMenu = () => {
		setMenuOpen(!isMenuOpen);
	};

	const closeMenu = () => {
		setMenuOpen(false);
	};

	return (
		<div className={`${props.className} menu`}>
			<ul className="menu__item">
				<li onClick={toggleMenu}>Game</li>
				{isMenuOpen && (
					<ul className="menu__submenu">
						<li className="menu__divider-bottom"
							onClick={() => {
								closeMenu();
								onNewGame();
							}}>New</li>
						{Object.values(levels).map((level) => {
							return (
								<li key={level.name}
									onClick={() => {
										closeMenu();
										onLevelChange(level)
									}}
									className={level.name === currentLevel.name ? 'menu__item--selected' : ''}>{level.name}</li>
							);
						})}
					</ul>
				)}
			</ul>
		</div>
	);
}

export default styled(menu)`
	background: var(--system-bg);
	box-sizing: border-box;
	height: 1.5rem;
	padding: 0.375rem;
	position: relative;
	font-size: 0;

	.menu__item {
		margin: 0;
		padding: 0;
		font-size: 0;
	}

	.menu__item > li {
		color: var(--system-text);
		cursor: default;
		display: inline-block;
		font-size: 0.75rem;
		line-height: 1;
		vertical-align: baseline;
	}

	.menu__submenu {
		background: var(--system-bg);
		border-bottom-color: var(--system-border-dark);
		border-left-color: var(--system-border-light);
		border-right-color: var(--system-border-dark);
		border-style: solid;
		border-top-color: var(--system-border-light);
		border-width: 1px;
		left: 0;
		margin: 0;
		min-width: 9rem;
		padding: 0;
		position: absolute;
		top: 100%;
	}
	.menu__submenu > li {
		box-sizing: border-box;
		cursor: default;
		font-size: 0.75rem;
		list-style-type: none;
		padding: 0.375rem 0.375rem 0.375rem 1.375rem;
		position: relative;
	}

	.menu__submenu > li.menu__item--selected::before {
		content: 'âœ“';
		display: block;
		position: absolute;
		top: 0.375rem;
		left: 0.375rem;
	}

	.menu__divider-bottom::after {
		content: '';
		display: block;
		position: absolute;
		top: 100%;
		left: 0.125rem;
		width: calc(100% - 0.25rem);
		height: 0;
		border-top: 0.0625rem var(--system-border-dark) solid;
		border-bottom: 0.0625rem var(--system-border-light) solid;
	}
`;
