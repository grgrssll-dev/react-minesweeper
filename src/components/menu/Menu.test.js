import React from 'react';
import { shallow } from 'enzyme/build';
import Menu from './Menu';
import Levels from '../../Levels';

const baseProps = {
	levels: Levels,
	currentLevel: Levels[0],
	onLevelChange: jest.fn(),
	onNewGame: jest.fn(),
	className: 'class-name',
};

describe('Menu [Unit]', () => {
	beforeEach(() => {
		jest.clearAllMocks();
	});

	it('should match component snapshot', () => {
		expect.assertions(1);
		const props = {
			...baseProps,
		};
		const wrapper = shallow(<Menu {...props} />);
		expect(wrapper).toMatchSnapshot();
	});
});