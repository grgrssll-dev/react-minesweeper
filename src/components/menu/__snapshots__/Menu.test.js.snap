// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu [Unit] should match component snapshot 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
